src/display/editor/highlight.js

```diff
@@ -0,0 +1,491 @@
+
+    return div;
+  }
+
+  pointerover() {
+    this.parent.drawLayer.addClass(this.#outlineId, "hovered");
+  }
+
+  pointerleave() {
+    this.parent.drawLayer.removeClass(this.#outlineId, "hovered");
+  }
+
+  /** @inheritdoc */
+  select() {
+    super.select();
+    this.parent?.drawLayer.removeClass(this.#outlineId, "hovered");
+    this.parent?.drawLayer.addClass(this.#outlineId, "selected");
+  }
+
+  /** @inheritdoc */
+  unselect() {
+    super.unselect();
+    this.parent?.drawLayer.removeClass(this.#outlineId, "selected");
+  }
+
+  #serializeBoxes() {
+    const [pageWidth, pageHeight] = this.pageDimensions;
+    const boxes = this.#boxes;
+    const quadPoints = new Array(boxes.length * 8);
+    let i = 0;
+    for (const { x, y, width, height } of boxes) {
+      const sx = x * pageWidth;
+      const sy = (1 - y - height) * pageHeight;
+      // The specifications say that the rectangle should start from the bottom
+      // left corner and go counter-clockwise.
+      // But when opening the file in Adobe Acrobat it appears that this isn't
+      // correct hence the 4th and 6th numbers are just swapped.
+      quadPoints[i] = quadPoints[i + 4] = sx;
+      quadPoints[i + 1] = quadPoints[i + 3] = sy;
+      quadPoints[i + 2] = quadPoints[i + 6] = sx + width * pageWidth;
+      quadPoints[i + 5] = quadPoints[i + 7] = sy + height * pageHeight;
+      i += 8;
+    }
+    return quadPoints;
+  }
+
+  #serializeOutlines() {
+    const [pageWidth, pageHeight] = this.pageDimensions;
+    const width = this.width * pageWidth;
+    const height = this.height * pageHeight;
+    const tx = this.x * pageWidth;
+    const ty = (1 - this.y - this.height) * pageHeight;
+    const outlines = [];
+    for (const outline of this.#highlightOutlines.outlines) {
+      const points = new Array(outline.length);
+      for (let i = 0; i < outline.length; i += 2) {
+        points[i] = tx + outline[i] * width;
+        points[i + 1] = ty + (1 - outline[i + 1]) * height;
+      }
+      outlines.push(points);
+    }
+    return outlines;
+  }
+
+  /** @inheritdoc */
+  static deserialize(data, parent, uiManager) {
+    const editor = super.deserialize(data, parent, uiManager);
+
+    const { rect, color, quadPoints } = data;
+    editor.#color = Util.makeHexColor(...color);
+    editor.#opacity = data.opacity;
+
+    const [pageWidth, pageHeight] = editor.pageDimensions;
+    editor.width = (rect[2] - rect[0]) / pageWidth;
+    editor.height = (rect[3] - rect[1]) / pageHeight;
+    const boxes = (editor.#boxes = []);
+    for (let i = 0; i < quadPoints.length; i += 8) {
+      boxes.push({
+        x: quadPoints[4] / pageWidth,
+        y: 1 - quadPoints[i + 5] / pageHeight,
+        width: (quadPoints[i + 2] - quadPoints[i]) / pageWidth,
+        height: (quadPoints[i + 5] - quadPoints[i + 1]) / pageHeight,
+      });
+    }
+    editor.#createOutlines();
+
+    return editor;
+  }
+
+  /** @inheritdoc */
+  serialize(isForCopying) {
```

Comment:
Nit: Maybe add a default value?
```suggestion
  serialize(isForCopying = false) {
```